pkg/statistics/handle/usage/indexusage/collector.go

```diff
@@ -0,0 +1,225 @@
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indexusage
+
+import (
+	"sync"
+	"time"
+
+	"github.com/pingcap/tidb/pkg/parser/model"
+	"github.com/pingcap/tidb/pkg/statistics/handle/usage/collector"
+)
+
+// GlobalIndexID is the key type for indexUsageMap.
+type GlobalIndexID struct {
+	TableID int64
+	IndexID int64
+}
+
+// Sample is the data structure to store index usage information.
+type Sample struct {
+	LastUsedAt     time.Time
+	QueryTotal     uint64
+	KvReqTotal     uint64
+	RowAccessTotal uint64
+	// 0, 0-1, 1-10, 10-20, 20-50, 50-100, 100
+	PercentageAccess [7]uint64
+}
+
+var bucketBound = [6]float64{0, 0.01, 0.1, 0.2, 0.5, 1.0}
+
+func getIndexUsageAccessBucket(percentage float64) int {
+	if percentage == 0 {
+		return 0
+	}
+
+	bucket := 0
+	for i := 1; i < len(bucketBound); i++ {
+		if percentage >= bucketBound[i-1] && percentage < bucketBound[i] {
+			bucket = i
+			break
+		}
+	}
+	if percentage == 1.0 {
+		bucket = len(bucketBound)
+	}
+
+	return bucket
+}
+
+// NewSample creates a new data point for index usage.
+func NewSample(queryTotal uint64, kvReqTotal uint64, rowAccess uint64, tableTotalRows uint64) Sample {
+	percentageAccess := [len(bucketBound) + 1]uint64{}
+
+	// if the `tableTotalRows` == 0, record the percentage as 1 and use the last bucket.
+	bucket := len(bucketBound)
+	if tableTotalRows > 0 {
+		rowAccessPercentage := float64(rowAccess) / float64(tableTotalRows)
+		bucket = getIndexUsageAccessBucket(rowAccessPercentage)
+	}
+	percentageAccess[bucket] = 1
+
+	return Sample{
+		QueryTotal:       queryTotal,
+		KvReqTotal:       kvReqTotal,
+		RowAccessTotal:   rowAccess,
+		PercentageAccess: percentageAccess,
+		LastUsedAt:       time.Now(),
+	}
+}
+
+type indexUsage map[GlobalIndexID]Sample
+
+func (m indexUsage) updateByKey(id GlobalIndexID, sample Sample) {
+	item := m[id]
+	item.QueryTotal += sample.QueryTotal
+	item.RowAccessTotal += sample.RowAccessTotal
+	item.KvReqTotal += sample.KvReqTotal
+	for i, val := range sample.PercentageAccess {
+		item.PercentageAccess[i] += val
+	}
+
+	if item.LastUsedAt.Before(sample.LastUsedAt) {
+		item.LastUsedAt = sample.LastUsedAt
+	}
+	m[id] = item
+}
+
+func (m indexUsage) update(tableID int64, indexID int64, sample Sample) {
+	id := GlobalIndexID{TableID: tableID, IndexID: indexID}
+	m.updateByKey(id, sample)
+}
+
+func (m indexUsage) merge(destMap indexUsage) {
+	for id := range destMap {
+		item := destMap[id]
+		m.updateByKey(id, item)
+	}
+}
+
+// Collector represents a data structure to record the index usage for the whole node
+type Collector struct {
+	collector collector.GlobalCollector[indexUsage]
+	mapper    indexUsage
```

Comment:
Could we rename `mapper` to another name like `indexUsage`?

Reply:
Done